Support --pid=container:xxx for `nerdctl run` cmd

minuk-dev opened the pull request:
#1293 
Signed-off-by: Min Uk Lee <minuk.dev@gmail.com>

### Changes
- support `--pid=container:xxx` for `nerdctl run` command
- extract `generatePIDOpts()` function for pid
- (Updated) update `setPlatformOptions()`
- (Updated) add `setPlatformContainerOptions()` - Details 1
- (Updated) add `PIDContainer` label - Details 2

### Details
#### 1. add `setPlatformContainerOptions()`
- Some functions only works on a specific platform (mostly linux). For persistency, they need to save information into container's labels.
- Previously, it was processed on runtime. (e.g. [hostname](https://github.com/containerd/nerdctl/blob/b4c01094b5815e7616c43151086c708ddd57bc37/cmd/nerdctl/run.go#L569), [logURI](https://github.com/containerd/nerdctl/blob/b4c01094b5815e7616c43151086c708ddd57bc37/cmd/nerdctl/run.go#L529)). It caused [too many number of `withInternalLabels()`'s parameters](https://github.com/containerd/nerdctl/blob/b4c01094b5815e7616c43151086c708ddd57bc37/cmd/nerdctl/run.go#L660)
- In my opinion, to keep up with it, if the number of retrun values ​​of `setPlatformOptions()` increase, it would be difficult to maintain the code quality.
- Therefore, I suggest adding `addPlatformContainerOptions()`.

#### 2. add `PIDContainer` label

> Thanks, but we need to deal with shared container restarts, it will change pid namesapce.

- For `--pid` flag's persistency, save the based container's id into the shared container's label.
- When the container restart, `reconfigPIDContainer()` will recover it in `startContainer()`

---
### a difference with docker
- Because docker supports to isolate a container with a user namespace, it should share user namespace between containers when containers share their pid namespace.
  - [docker docs](https://docs.docker.com/engine/security/userns-remap/)
  - [moby/moby/daemon/oci_linux.go#L231-L240](https://github.com/moby/moby/blob/6c1df4f9c53b604866526e0f2aa4d34614c5f7ab/daemon/oci_linux.go#L231-L240)
- But, nerdctl does not support the user ns isolation.

### references
- [moby/moby/daemon/oci_linux.go#L294-L321](https://github.com/moby/moby/blob/6c1df4f9c53b604866526e0f2aa4d34614c5f7ab/daemon/oci_linux.go#L294-L321)
- [man pages - pid_namespaces](https://man7.org/linux/man-pages/man7/pid_namespaces.7.html)

junnplus commented on cmd/nerdctl/run_linux_test.go:
```diff
@@ -75,6 +75,18 @@ func TestRunPidHost(t *testing.T) {
+	t.Parallel()
+	base := testutil.NewBase(t)
+
+	sharedContainerResult := base.Cmd("run", "-d", testutil.AlpineImage, "sleep", "30").Run()
```
sleep infinity to avoid flaky.

junnplus commented on cmd/nerdctl/run_linux.go:
```diff
@@ -217,3 +210,90 @@ func withOOMScoreAdj(score int) oci.SpecOpts {
+					return fmt.Errorf("multiple IDs found with provided prefix: %s", found.Req)
+				}
+
+				nc, err := containerinspector.Inspect(ctx, found.Container)
+				if err != nil {
+					return err
+				}
```
Avoid using inspect, it will do extra work

junnplus commented on cmd/nerdctl/run_linux.go:
```diff
@@ -217,3 +210,90 @@ func withOOMScoreAdj(score int) oci.SpecOpts {
+		}
+	}
+
+	return oci.Compose(opts...), nil
```
why compose?

minuk-dev replied:
no need. I'll fix the function returns []oci.Specs.

junnplus reviewed (changes requested):
We need to deal with shared container restarts, it will change pid namesapce.

And please update the doc.

minuk-dev commented:
> Thanks, but we need to deal with shared container restarts, it will change pid namesapce.

Could I get an usecase for it? I could not find it because of my little knowledge.
In my opinion, if containers restart, their namespaces should be reset, isn't it?

Here are my manual tests.
In these cases, docker and nerdctl work the same.

- restart base container test
```bash
$ docker run -d ubuntu sleep infinity # run a base container
9bd0d57b85dcbbe7eeabf72f867bb130ddaf35ca716b80f15f6612b92705124f

$ docker run --pid=container:9bd0d -d ubuntu sleep infinity # run shared container
cfe162ae92814624e44f06db81ecba3f016508a72708ee917ad9e60fe0fe663c

$ docker ps # check containers' status
CONTAINER ID   IMAGE     COMMAND            CREATED          STATUS          PORTS     NAMES
cfe162ae9281   ubuntu    "sleep infinity"   7 seconds ago    Up 6 seconds              gallant_galois
9bd0d57b85dc   ubuntu    "sleep infinity"   45 seconds ago   Up 44 seconds             peaceful_mendel

$ docker restart 9bd0d # restart the base container
9bd0d

$ docker ps # only base container survived
CONTAINER ID   IMAGE     COMMAND            CREATED              STATUS        PORTS     NAMES
9bd0d57b85dc   ubuntu    "sleep infinity"   About a minute ago   Up 1 second             peaceful_mendel
```

- restart shared container test
```
$ docker run -d ubuntu sleep infinity # run a base container
57f18c82ab5ebbf49d6adbded478abf85d71172ee06de5e0fb88acf552493957
$ docker run --pid=container:57f18 -d ubuntu sleep infinity # run a shared container
9de0f487563cb709f0065aecc6a02c4eccb2c2c5089ad5bcf885250ac993a4a1

$ docker restart 9de0f # restart the shared container
9de0f

$ docker ps 
CONTAINER ID   IMAGE     COMMAND            CREATED          STATUS         PORTS     NAMES
9de0f487563c   ubuntu    "sleep infinity"   16 seconds ago   Up 2 seconds             gracious_dewdney
57f18c82ab5e   ubuntu    "sleep infinity"   2 minutes ago    Up 2 minutes             beautiful_almeida

# all survived but, not shared.
$ docker exec -it 9de0f bash
root@9de0f487563c:/# ps
    PID TTY          TIME CMD
     18 pts/0    00:00:00 bash
     26 pts/0    00:00:00 ps
```

fahedouch commented on cmd/nerdctl/run_linux.go:
```diff
@@ -217,3 +209,94 @@ func withOOMScoreAdj(score int) oci.SpecOpts {
+			opts = append(opts, withBindMountHostProcfs)
+		}
+	default: // container:<id|name>
+		client, ctx, cancel, err := newClient(cmd)
```
platform different from linux signature should be like this 
```
func setPlatformOptions(opts []oci.SpecOpts, cmd *cobra.Command, client *containerd.Client, id string) ([]oci.SpecOpts, error) {
      return opts, nil
} 
```

minuk-dev replied:
Thanks for your kind reply.
I updated it on 8de12b934788734ce480cdc2d2b8e338d2af3697.

minuk-dev commented:
> > Could I get an usecase for it? I could not find it because of my little knowledge.
> > In my opinion, if containers restart, their namespaces should be reset, isn't it?
> 
> You can test restart the base container and then restart the shared container.

I fixed it using labels at 7a6d4250619c76eda7d2d2d748d9cf2fbc24313e.

Summary of updates:
- add `setPlatformContainerOptions()` at 82e3d04723bc3bd86f7dd075da5ff091e6432022
- add `reconfigPIDContainer()`
- add `PIDContainer` label
- test this feature in [TestRestartPIDContainer](https://github.com/minuk-dev/nerdctl/blob/7a6d4250619c76eda7d2d2d748d9cf2fbc24313e/cmd/nerdctl/restart_linux_test.go#L48).

Can you review it again? @junnplus 

trivial: I updated the doc of pr. It is marked as updated.

junnplus commented on cmd/nerdctl/restart_linux_test.go:
```diff
@@ -42,3 +44,28 @@ func TestRestart(t *testing.T) {
+	base := testutil.NewBase(t)
+
+	baseContainerResult := base.Cmd("run", "-d", testutil.AlpineImage, "sleep", "infinity").Run()
+	baseContainerID := strings.TrimSpace(baseContainerResult.Stdout())
```
Using `testutil.Identifier(t)` as container name.

minuk-dev replied:
I know that the run command with `-d` should print a container's id.
Why should we name for it?

junnplus replied:
Ask in turn, why do we need to parse the id from stdout?

junnplus commented on cmd/nerdctl/run_linux.go:
```diff
@@ -217,3 +229,85 @@ func withOOMScoreAdj(score int) oci.SpecOpts {
+			return nil, fmt.Errorf("no such container: %s", containerName)
+		}
+
+		o, err := generateSharingPIDOpts(ctx, targetCon)
+		if err != nil {
+			return nil, err
+		}
+
+		opts = append(opts, o...)
```
Handle in `OnFound`

junnplus commented on cmd/nerdctl/run_linux.go:
```diff
@@ -217,3 +229,85 @@ func withOOMScoreAdj(score int) oci.SpecOpts {
+	}
+
+	containerName := parsed[1]
+	walker := &containerwalker.ContainerWalker{
```
Why do we need to find again?

junnplus commented on cmd/nerdctl/run_linux.go:
```diff
@@ -217,3 +229,85 @@ func withOOMScoreAdj(score int) oci.SpecOpts {
+			if found.MatchCount > 1 {
+				return fmt.Errorf("multiple IDs found with provided prefix: %s", found.Req)
+			}
+			cOpts = append(cOpts, containerd.WithAdditionalContainerLabels(map[string]string{
+				labels.PIDContainer: containerName,
+			}))
```
We already have `withInternalLabels`, should be put together

minuk-dev replied:
When a new label is needed, the best way is to add a new parameter of `withInternalLabels()`?
And if so, `--pid` flags is processed in `setPlatformOptions()`. In order to pass into the `withInternalLabels()`, the number of `setPlatformOptions()`'s return value should be increased.

junnplus replied:
> And if so, --pid flags is processed in setPlatformOptions().

I think the `--pid` flags should be platform independent, why not consider moving it out of `setPlatformOptions`?

minuk-dev replied:
Previsouly, `--pid` only works on linux.
https://github.com/containerd/nerdctl/blob/c6517c7f9a178950b965748f6698f5ca4e76e318/cmd/nerdctl/run_linux.go#L135-L144

The reason, why it only works on the linux, is that it's too sensitive with namespace suppport.

https://github.com/containerd/nerdctl/blob/d15ada7bec3df3a4e2b6d78bdc9f5d518aaf9d44/cmd/nerdctl/run.go#L1184-L1187

```golang
// WithHostNamespace allows a task to run inside the host's linux namespace
func WithHostNamespace(ns specs.LinuxNamespaceType) SpecOpts {
	return func(_ context.Context, _ Client, _ *containers.Container, s *Spec) error {
		setLinux(s)
		for i, n := range s.Linux.Namespaces {
			if n.Type == ns {
				s.Linux.Namespaces = append(s.Linux.Namespaces[:i], s.Linux.Namespaces[i+1:]...)
				return nil
			}
		}
		return nil
	}
}
```

To avoid a compile error, I placed `reconfigPIDContainer()` in `cmd/nerdctl/start.go` but I think it caused confusing.
So, I added a condition to check runtime os now. https://github.com/containerd/nerdctl/blob/8680bdb9a24d0c99fba6682af7473124469cb2fd/cmd/nerdctl/start.go#L231

Anyway, in my opinion, `--pid` currently can work on linux only.
Because of it, `--pid` should be processed in `setPlatformOptions()`. But, I really agree with you. `setPlatformContainerOptions()` can make a fragmentation and cause to find a target container twice.
I tried to find a better way to improve it according to your review, but I couldn't. I'm really sorry.

junnplus reviewed (changes requested):
Please squash commits

AkihiroSuda reviewed (approved):
Thanks

minuk-dev commented:
> Thanks

I'm sorry that I didn't apply reviews from jun. I'm looking forward the better way than setplatformcontaineropts().
In my opinion, that's too hard to me now.
So, could you decide how to deal with this pr (e.g. because of low code quality, just close).
If pr is closed, then I'll try to improve the quality and retry pr.

junnplus reviewed (changes requested):
Sorry, I guess the code can be optimized again.

AkihiroSuda commented:
@junnplus LGTY?

AkihiroSuda commented:
Let me merge this, as I don't think the `--pid` flag can be platform-independent in the foreseeable future, but please feel free to submit follow-up issues/PRs.

Thank you @minuk-dev @junnplus 👍
